### ISSUE TITLE
Python: DMatrix.slice() looses feature_names

### ISSUE THREAD

**Issue by tarstars** [dmlc/xgboost#3124]
One can see that DMatrix.split operation drops feature_names to default:

```
def demonstrate_split_problem():
    pool = xgboost.DMatrix(np.array([[0], [1]]), feature_names=['a'])
    subpool = pool.slice([0])
    if pool.feature_names != subpool.feature_names:
        print('I told you so!')
```

It is due to DMatrix.__init__() issue:
``` #first lines of DMatrix.__init__:
          if data is None:  # skips rest of the function in case of None input data
            self.handle = None
            return
  #end of the function:
  self.feature_names = feature_names
  self.feature_types = feature_types
```

But DMatrix.slice() use just this scenario of DMatrix.__init__() usage:

```
res = DMatrix(None, feature_names=self.feature_names)  # here feature_names is lost
```

I propose to move 
```  self.feature_names = feature_names ```
line to the end of the ```split``` function.
  

## Steps to reproduce

1. Create DMatrix with feature_names
2. call split()
3. check current feature_names state

**Comment by tqchen** [comment]
This seems to be an issue that we can fix, can you send a PR to fix this?

**Comment by hcho3** [comment]
Consolidating to #3439. A new issue should be opened if someone decides to actively work on implementing this feature.

**Comment by ma3axaka** [comment]
@hcho3 Could I work on this one? Do I need to reopen it?

**Comment by trivialfis** [comment]
@ma3axaka Please open a new issue.  We want to do some refactorings around DMatrix.  So maybe we can work on it together.

**Comment by hcho3** [comment]
@trivialfis Can you open a new RFC about the proposed re-factoring? (Doesn't have to be too elaborate; just explain what you want to achieve)

**Comment by trivialfis** [comment]
@RAMitchell Here.

**Comment by ma3axaka** [comment]
@trivialfis Is there a ticket for refactoring DMatrix?

**Comment by trivialfis** [comment]
@ma3axaka Not yet.  Should be up next week, we want to summarize collected ideas, thanks for the patience. ;)

**Comment by hcho3** [comment]
@tarstars @bourbaki Fixed by #3766
